subprojects/core-api/src/main/java/org/gradle/api/artifacts/ConfigurationVariant.java

```diff
@@ -17,17 +17,28 @@
 package org.gradle.api.artifacts;
 
 import org.gradle.api.Action;
+import org.gradle.api.Incubating;
 import org.gradle.api.Named;
 import org.gradle.api.attributes.HasConfigurableAttributes;
 import org.gradle.internal.HasInternalProtocol;
 
+import javax.annotation.Nullable;
+
 /**
  * Represents some variant of an outgoing configuration.
  *
  * @since 3.3
  */
 @HasInternalProtocol
 public interface ConfigurationVariant extends Named, HasConfigurableAttributes<ConfigurationVariant> {
+    /**
+     * Returns an optional note describing this variant.
+     *
+     * @since 7.5
+     */
+    @Incubating
+    @Nullable
+    String getDescription();
```

Comment:
The problem is that secondary variants aren't created as `Configuration`.  They are only available as `ConfigurationVariant`, retrieved via `configuration.getOutgoing().getVariants()`, and created as in `DefaultJvmPluginServices#registerClassesDirVariant`, or `JavaPlugin#addRuntimeVariants`.  Currently there is no way to put any sort of note on them describing how they are meant to be used.

We thought this would be a helpful addition, especially if we want to move in the direction of bringing top-level configurations and variants closer, possibly eventually re-uniting them.  It might help demystify the reasons for some of these things existing.

In `DefaultVariant` I implemented this by using the `Describable` already present (this could be a possible default?), which currently reports 'configuration ':apiElements' variant classes'.  Both this class and `DefaultConfigurationVariant` would have to pick up a new desc field and setters/additional constructor args to make the new description possible.

We could alternately set up some sort of registry where these variants are mapped to descriptions, and then read that at reporting time, but this seemed simpler - hopefully a new `@Nullable` getter with an obvious purpose won't cause many issues.

Reply:
Let's add it